### PR TITLE
pkg/p2p: ackManager for implementing MessageServer

### PR DESCRIPTION
**Pull request by liuzix**
### What problem does this PR solve? 
- We need a thread-safe data structure to record the progress of peer-message handling.

Screenshot from 2021-11-22 16-57-18

This PR implements a data structure managing the `acks` in this picture.

### What is changed and how it works?
- Implemented `ackManager`.

### Check List 

Tests 

 - Unit test

### Release note 

```release-note
None
```

### CODE REVIEW

File: pkg/p2p/server_ack_manager.go [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+}
+
+// Set sets the latest ACK for a given topic sent from a given node.
+func (m *ackManager) Set(senderID NodeID, topic Topic, newSeq Seq) {
+	rawAcks, ok := m.peers.Load(senderID)
+	if !ok {
+		newAcks := &peerAckList{
+			acks: make(map[Topic]Seq),
+		}
+		rawAcks, _ = m.peers.LoadOrStore(senderID, newAcks)
```
Comment (overvenus) [inline pkg/p2p/server_ack_manager.go:64]: What if it already exists?

File: pkg/p2p/server_ack_manager.go [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+// newAckManager returns a new ackManager
+func newAckManager() *ackManager {
+	return &ackManager{}
+}
+
+// Get returns the latest ACK for a given topic sent from a given node.
+func (m *ackManager) Get(senderID NodeID, topic Topic) Seq {
+	rawAcks, ok := m.peers.Load(senderID)
+	if !ok {
+		return 0
```
Comment (amyangfei) [inline pkg/p2p/server_ack_manager.go:47]: Can add a const value for special Seq.

File: pkg/p2p/server_ack_manager.go [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+		newAcks := &peerAckList{
+			acks: make(map[Topic]Seq),
+		}
+		rawAcks, _ = m.peers.LoadOrStore(senderID, newAcks)
+	}
+
+	ackList := rawAcks.(*peerAckList)
+	ackList.mu.Lock()
+	defer ackList.mu.Unlock()
+	ackList.acks[topic] = newSeq
```
Comment (amyangfei) [inline pkg/p2p/server_ack_manager.go:70]: Does `ackManager` allow `Seq` fallen back here?

Comment (liuzix) [reply pkg/p2p/server_ack_manager.go:70]: It does not care.

**Review by asddongmen — APPROVED** [review]

**Review by amyangfei — APPROVED** [review]

**Pull request merged**